src/librbd/migration/NBDStream.cc

```diff
@@ -0,0 +1,263 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/migration/NBDStream.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/AsioEngine.h"
+#include "librbd/ImageCtx.h"
+
+#include <libnbd.h>
+
+namespace librbd {
+namespace migration {
+
+namespace {
+
+const std::string SERVER_KEY {"server"};
+const std::string PORT_KEY {"port"};
+
+int extent_cb(void *data, const char *metacontext, uint64_t offset,
+              uint32_t *entries, size_t nr_entries, int *error) {
+  auto sparse_extents = reinterpret_cast<io::SparseExtents*>(data);
+
+  uint64_t length = 0;
+  for (size_t i=0; i<nr_entries; i+=2) {
+    length += entries[i];
+  }
+  auto state = io::SPARSE_EXTENT_STATE_DATA;
+  if (nr_entries == 2) {
+    if (entries[1] & (LIBNBD_STATE_HOLE | LIBNBD_STATE_ZERO)) {
+      state = io::SPARSE_EXTENT_STATE_ZEROED;
+    }
+  }
+  sparse_extents->insert(offset, length, {state, length});
+  return 1;
+}
+
+} // anonymous namespace
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::migration::NBDStream::ReadRequest: " \
+                           << this << " " << __func__ << ": "
+
+template <typename I>
+struct NBDStream<I>::ReadRequest {
+  NBDStream* nbd_stream;
+  io::Extents byte_extents;
+  bufferlist* data;
+  Context* on_finish;
+
+  ReadRequest(NBDStream* nbd_stream, io::Extents&& byte_extents,
+              bufferlist* data, Context* on_finish)
+    : nbd_stream(nbd_stream), byte_extents(std::move(byte_extents)),
+      data(data), on_finish(on_finish) {
+  }
+
+  void send() {
+    data->clear();
+    read();
+  }
+
+  void read() {
+    auto cct = nbd_stream->m_cct;
+    ldout(cct, 20) << "byte_extents=" << byte_extents << dendl;
+
+    for (auto [byte_offset, byte_length] : byte_extents) {
+      ldout(cct, 20) << "byte_offset=" << byte_offset << " byte_length="
+                     << byte_length << dendl;
+      auto ptr = buffer::ptr_node::create(buffer::create_small_page_aligned(
+        byte_length));
+      int rc = nbd_pread(nbd_stream->m_nbd, ptr->c_str(), byte_length,
+                         byte_offset, 0);
+      data->push_back(std::move(ptr));
```

Comment:
Ideally it would be done right after creating `ptr`, but the problem is that after `std::move(ptr)` is performed passing the buffer to `nbd_pread()` with `ptr->c_str()` wouldn't work.  So it's just done as close to the line that creates `ptr` as possible -- `rc` value of the call is irrelevant.

Reply:
Reading https://libguestfs.org/nbd_pread.3.html , `nbd_pread` reads data into buffer pointed by `ptr` . When `rc` is -1 doesn't that indicate that buffer pointed by `ptr` may not be valid? If so, then why do we push back the buffer into the buffer list `data` irrespective of whether `nbd_pread()` succeeds or not?

Reply:
> When `rc` is -1 doesn't that indicate that buffer pointed by `ptr` may not be valid?

Yes, but `finish()` clears the bufferlist on errors (and if it didn't, the caller would be expected to not trust it anyway).

> If so, then why do we push back the buffer into the buffer list `data` irrespective of whether `nbd_pread()` succeeds or not?

No good reason.  The original code code was more convoluted and involved wrapping `ptr` with `boost::asio::mutable_buffer()` only to unwrap it a couple of lines down -- completely bogus, of course.  I made a minimal change to clean that up and it ended up like this.

I'll move the `push_back()` call to be performed only in case `nbd_pread()` succeeds.